spec/src/main/asciidoc/microprofile-lra-spec.adoc

```diff
@@ -1009,26 +1009,36 @@ example, to indicate that an LRA should automatically cancel after 100 milliseco
       @HeaderParam(LRA.LRA_HTTP_CONTEXT_HEADER) URI lraId) {...}
 ----
 
-Furthermore, the ability to compensate or complete may be transient capabilities of a
-service so participants can also be timed out. When the time limit is reached
-the LRA is cancelled and participants will be notified via their compensation
-(i.e. the method annotated with `@Compensate`). To set such a time limit add the
-`timeLimit` and `timeUnit` element to the `@Compensate` and `@Complete`
-annotations. If different time limits are set on these two methods then the
-earliest of the two will take effect. An example of how a participant could
-indicate that its' ability to compensate is limited to 100 milliseconds could be:
+When the timeout is reached, the LRA moves to the state `Cancelling` and all registered participants receive a `Compensate`
+call. Just as if the Long Running Action would be closed due to an error.
+
+When a participant joins an existing transaction using the `@LRA(value = ...)` annotation, the timeout of
+the already active Long Running Action can be influenced by the `timeLimit` and `timeUnit` defined on this annotation.
+The implementation should calculate the moment in time the LRAs would become eligible for cancellation based on the time
+the request enters the JAX-RS method and the timeout information found on the `@LRA`.  When this moment is earlier than
+the moment calculated for the LRA at that time, this new moment becomes the time out moment for the LRA.
+
+An example to illustrate this, suppose you have the following endpoint definitions
 
 [source,java]
 ----
-    @Compensate(timeLimit = 100, timeUnit = ChronoUnit.MILLIS)
-    @Path("/compensate")
-    @PUT
-    public Response completeWork(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId,
-                                 String userData) { ... }
+  @LRA(timeLimit = 1, timeUnit = ChronoUnit.HOURS, end = false)
+  @Path("/someService")
+  @GET
+  public Response someService(...) {...}
+
+  @LRA(value = LRA.Type.MANDATORY, timeLimit = 1, timeUnit = ChronoUnit.MINUTES, end = false)
+  @Path("/anotherService")
+  @GET
+  public Response anotherService(...) {...}
+
 ----
 
-In this example after 100 milliseconds has passed the implementation SHOULD
-automatically cancel the LRA which will result in this method being invoked.
+Suppose you have a call to the `someService` endpoint at 10h25, the time out is at that moment calculated at 11h25.
+Suppose that `anotherService` is called for the same Long Running Action, this is the LRAid is in the header for
+the call at the endpoint, at 10h27. The timeout related to this call is 10h28. Since these 2 calls share the same
+Long Running action, the LRA will be eligible for automatic cancellation from 10h28 on although the time out was initially 1 hour,
+it can be cancelled already after 3 minutes.
```

Comment:
This text seems to imply that services are using synchronised clocks. I would just say that if two timeout values have been set on an active context then the earliest one that reaches the implementation will trigger cancellation of the LRA.

Reply:
reworded

Reply:
The text is still using absolute times and would require the two services to have to synchronise their clocks. The impossibility of synchronising clocks across entities in a distributed system was discussed in one of Lamport's papers. I would just say that if two timeout values have been set on an active context then the earliest one calculated by the implementation will trigger cancellation of the LRA.

Reply:
For example if the message that carries the timeout values is delayed in the network then this would throw off the absolute time calculations.

Reply:
> if two timeout values have been set on an active context then the earliest one calculated by the implementation will trigger cancellation of the LRA.

This also assumes synchronization as it needs to know the 'earliest one'

Reply:
So if we do not have any possibility to reliably determine the time, then we should change the spec and indicate that defining another timeout on subsequent participants is forbidden.

Reply:
All posix compliant clocks are monotonic so the concept of earliest does not depend on synchronised clocks. A clock which ticks backwards would brake all kinds of stuff. Therefore a local clock can reliably order time. 

If the implementation is distributed across nodes then the first node whose time scheduler fires will trigger cancellation of the LRA.

Reply:
Let me elaborate more on what I meant (I was at a conference and replied too condensed)

The example I have put in the spec is from the point of view of the end user. He has 'absolute' timings, namely from his perspective, there are things happening on the services as a response to his actions.
So when he is watching his clock, a request can hit service A at 10h25, and Service B at 10h27. And this regardless of the timings on the Servers running Service A and B.

And the 'earliest' still assumes some kind of synchronization of the clocks.  200 years from now in the western calendar (say the year 2220) is already a few hundred years in the past according to the Buddhist Calendar today.

Clocks can go back in time! When servers are using the time Services, adjustments can happen so that clock jumps back. And even a more obvious example, when we switch from Summer to Winter time, clocks go back 1 hour.
 
So this discussion has highlighted some other cases we need to explain in the specification.
- The timeouts are durations as you mention, but they result in some calculated end time. In relation to which system do they get calculated? (the participant or in case there is a central component)
- In which frame of reference are those 'end times' calculated. A difference in timezone (but also using the Buddhist Calendar for example) will not result in the expected outcome. And what to do with the switches from Winter to Summer Time and vice versa.
We need to mention that, although an absolute synchronization between clocks is not possible, they need to be 'aligned'. When a participant wants to join a transaction but the clock has a difference with the other services, it is possible that joining will fail as according to this service the LRA is already timed-out.

I'll create another ticket for the above 'timing' issues.

Reply:
What if I add a sentence that the clock values are from the users perspective and that no absolute clock  synchronization is assumed between the services?

Would that be OK?

Reply:
Not really. You are adding complexity where none is necessary.

A service joins and a timer is started. Similarly on a second server. Here it is irrelevant whether or not two clocks read different times since the implementation just needs to wait for a period to elapse and the two periods should be identical since all practical clocks tick at the same rate.

The only circumstance where we would ever need to use a time base is for crash recovery purposes and the use of UTC time would side step any concerns over timezones.

What is wrong with the simple short sentence that I suggested?

Reply:
I don't think a link was provided here so the ticket that Rudy created is this one I think: https://github.com/eclipse/microprofile-lra/issues/237 (I added a comment on there)

Reply:
>What is wrong with the simple short sentence that I suggested?

It doesn't make it very clear that although a participant has defined a timeout, it can be overruled when within the same LRA another endpoint is called with another timeout.

But removed example and absolute timings and added the simple sentence.